File: src/Application Specific/Classes/Entity Collections/Warehouse/WarehouseSQL.vb
Imports System.Data.SqlClient

Namespace Entity

    Namespace Warehouses

        Public Class WarehouseSQL
            Private _database As Sys.Database

            Public Sub New(ByVal database As Sys.Database)
                _database = database
            End Sub

#Region "Functions"

            Public Sub Delete(ByVal WarehouseID As Integer)
                Dim WarehouseName As String = Warehouse_Get(WarehouseID).Name.Trim

                _database.ClearParameter()
                _database.AddParameter("@WarehouseID", WarehouseID, True)
                _database.ExecuteSP_NO_Return("Warehouse_Delete")

                For Each row As DataRow In DB.Van.Van_GetAll(False).Table.Rows
                    If Sys.Helper.MakeStringValid(row.Item("Registration")).Split("*")(1).Trim = WarehouseName Then
                        DB.Van.Delete(row.Item("VanID"), True)
                    End If
                Next
            End Sub

            Public Function [Warehouse_GetDV](ByVal WarehouseID As Integer, ByVal trans As SqlClient.SqlTransaction) As DataView

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_Get"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@WarehouseID", WarehouseID)


                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)

            End Function

            Public Function [Warehouse_GetDV](ByVal WarehouseID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@WarehouseID", WarehouseID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_Get")
                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)

            End Function

            Public Function [Warehouse_Get](ByVal WarehouseID As Integer) As Warehouse

                _database.ClearParameter()
                _database.AddParameter("@WarehouseID", WarehouseID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_Get")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oWarehouse As New Warehouse
                        With oWarehouse
                            .IgnoreExceptionsOnSetMethods = True
                            .SetWarehouseID = dt.Rows(0).Item("WarehouseID")
                            .SetName = dt.Rows(0).Item("warehouseName")
                            .SetSize = dt.Rows(0).Item("Size")
                            .SetNotes = dt.Rows(0).Item("Notes")
                            .SetAddress1 = dt.Rows(0).Item("Address1")
                            .SetAddress2 = dt.Rows(0).Item("Address2")
                            .SetAddress3 = dt.Rows(0).Item("Address3")
                            .SetAddress4 = dt.Rows(0).Item("Address4")
                            .SetAddress5 = dt.Rows(0).Item("Address5")
                            .SetPostcode = dt.Rows(0).Item("Postcode")
                            .SetTelephoneNum = dt.Rows(0).Item("TelephoneNum")
                            .SetFaxNum = dt.Rows(0).Item("FaxNum")
                            .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            .SetActive = dt.Rows(0).Item("Active")

                            .SetDeleted = dt.Rows(0).Item("Deleted")
                        End With
                        oWarehouse.Exists = True
                        Return oWarehouse
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [Warehouse_Get](ByVal WarehouseID As Integer, ByVal trans As SqlClient.SqlTransaction) As Warehouse

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_Get"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@WarehouseID", WarehouseID)
             

                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oWarehouse As New Warehouse
                        With oWarehouse
                            .IgnoreExceptionsOnSetMethods = True
                            .SetWarehouseID = dt.Rows(0).Item("WarehouseID")
                            .SetName = dt.Rows(0).Item("warehouseName")
                            .SetSize = dt.Rows(0).Item("Size")
                            .SetNotes = dt.Rows(0).Item("Notes")
                            .SetAddress1 = dt.Rows(0).Item("Address1")
                            .SetAddress2 = dt.Rows(0).Item("Address2")
                            .SetAddress3 = dt.Rows(0).Item("Address3")
                            .SetAddress4 = dt.Rows(0).Item("Address4")
                            .SetAddress5 = dt.Rows(0).Item("Address5")
                            .SetPostcode = dt.Rows(0).Item("Postcode")
                            .SetTelephoneNum = dt.Rows(0).Item("TelephoneNum")
                            .SetFaxNum = dt.Rows(0).Item("FaxNum")
                            .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            .SetActive = dt.Rows(0).Item("Active")

                            .SetDeleted = dt.Rows(0).Item("Deleted")
                        End With
                        oWarehouse.Exists = True
                        Return oWarehouse
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [Warehouse_GetByLocationID](ByVal LocationID As Integer) As Warehouse
                _database.ClearParameter()
                _database.AddParameter("@LocationID", LocationID)

                'Get the datatable from the database store in dt
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_GetByLocationID")

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oWarehouse As New Warehouse
                        With oWarehouse
                            .IgnoreExceptionsOnSetMethods = True
                            .SetWarehouseID = dt.Rows(0).Item("WarehouseID")
                            .SetName = dt.Rows(0).Item("warehouseName")
                            .SetSize = dt.Rows(0).Item("Size")
                            .SetNotes = dt.Rows(0).Item("Notes")
                            .SetAddress1 = dt.Rows(0).Item("Address1")
                            .SetAddress2 = dt.Rows(0).Item("Address2")
                            .SetAddress3 = dt.Rows(0).Item("Address3")
                            .SetAddress4 = dt.Rows(0).Item("Address4")
                            .SetAddress5 = dt.Rows(0).Item("Address5")
                            .SetPostcode = dt.Rows(0).Item("Postcode")
                            .SetTelephoneNum = dt.Rows(0).Item("TelephoneNum")
                            .SetFaxNum = dt.Rows(0).Item("FaxNum")
                            .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            .SetActive = dt.Rows(0).Item("Active")

                            .SetDeleted = dt.Rows(0).Item("Deleted")
                        End With
                        oWarehouse.Exists = True
                        Return oWarehouse
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [Warehouse_GetByLocationID](ByVal LocationID As Integer, ByVal trans As SqlClient.SqlTransaction) As Warehouse

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_GetByLocationID"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@LocationID", LocationID)


                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                If Not dt Is Nothing Then
                    If dt.Rows.Count > 0 Then
                        Dim oWarehouse As New Warehouse
                        With oWarehouse
                            .IgnoreExceptionsOnSetMethods = True
                            .SetWarehouseID = dt.Rows(0).Item("WarehouseID")
                            .SetName = dt.Rows(0).Item("warehouseName")
                            .SetSize = dt.Rows(0).Item("Size")
                            .SetNotes = dt.Rows(0).Item("Notes")
                            .SetAddress1 = dt.Rows(0).Item("Address1")
                            .SetAddress2 = dt.Rows(0).Item("Address2")
                            .SetAddress3 = dt.Rows(0).Item("Address3")
                            .SetAddress4 = dt.Rows(0).Item("Address4")
                            .SetAddress5 = dt.Rows(0).Item("Address5")
                            .SetPostcode = dt.Rows(0).Item("Postcode")
                            .SetTelephoneNum = dt.Rows(0).Item("TelephoneNum")
                            .SetFaxNum = dt.Rows(0).Item("FaxNum")
                            .SetEmailAddress = dt.Rows(0).Item("EmailAddress")
                            .SetActive = dt.Rows(0).Item("Active")

                            .SetDeleted = dt.Rows(0).Item("Deleted")
                        End With
                        oWarehouse.Exists = True
                        Return oWarehouse
                    Else
                        Return Nothing
                    End If
                Else
                    Return Nothing
                End If
            End Function

            Public Function [Warehouse_GetAll]() As DataView
                _database.ClearParameter()
                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_GetAll")
                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)
            End Function

            Public Function Warehouse_GetAll_For_Van2(ByVal VanID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@VanID", VanID)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_GetAll_For_Van2")  ' rd change
                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)
            End Function

            Public Function [Warehouse_GetAll](ByVal trans As SqlClient.SqlTransaction) As DataView

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_GetAll"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection


                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)
                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)

            End Function

            Public Function [Insert](ByVal oWarehouse As Warehouse, ByVal LocationsDataView As DataView) As Warehouse
                _database.ClearParameter()
                AddWarehouseParametersToCommand(oWarehouse)

                oWarehouse.SetWarehouseID = Entity.Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Warehouse_Insert"))
                oWarehouse.Exists = True

                Dim oLocation As New Entity.Locationss.Locations
                oLocation.SetDeleted = False
                oLocation.SetVanID = Nothing
                oLocation.SetWarehouseID = oWarehouse.WarehouseID
                DB.Location.Insert(oLocation)

                For Each row As DataRow In DB.Van.Van_GetAll(True).Table.Rows
                    Dim oVan As New Vans.Van
                    oVan.SetRegistration = row.Item("Registration") & " * " & oWarehouse.Name

                    oVan.SetNotes = row.Item("Notes")
                    oVan.InsuranceDue = row.Item("InsuranceDue")
                    oVan.MOTDue = row.Item("MOTDue")
                    oVan.TaxDue = row.Item("TaxDue")
                    oVan.ServiceDue = row.Item("ServiceDue")
                    oVan.SetMileageLimit = row.Item("MileageLimit")
                    oVan.SetPeriodValue = row.Item("PeriodValue")
                    oVan.SetPeriodType = row.Item("PeriodType")

                    DB.Van.Insert(oVan, "", Nothing, True)
                Next

                For Each row As DataRow In DB.Van.Van_GetAll(True).Table.Rows
                    For Each locationRow As DataRow In LocationsDataView.Table.Rows
                        If row.Item("Registration") = locationRow.Item("Registration") Then
                            Dim locID As Integer = DB.Van.Van_GetAll_For_Warehouse(oWarehouse.WarehouseID).Table.Select("Registration = '" & row.Item("Registration") & "'")(0).Item("LocationID")
                            DB.Location.Update(locID, Not locationRow.Item("Tick"))
                        End If
                    Next
                Next

                Return oWarehouse
            End Function

            Public Function [Warehouse_Search](ByVal criteria As String, ByVal SearchOnField As String) As DataView

                If SearchOnField.Trim.Length > 0 Then
                    criteria = "'%" & criteria & "%'"
                End If

                _database.ClearParameter()
                _database.AddParameter("@SearchString", criteria, True)
                _database.AddParameter("@SearchOnField", SearchOnField, True)

                Dim dt As DataTable = _database.ExecuteSP_DataTable("Warehouse_Search")
                dt.TableName = Entity.Sys.Enums.TableNames.tblWarehouse.ToString
                Return New DataView(dt)
            End Function

            Public Sub [Update](ByVal oWarehouse As Warehouse, ByVal LocationsDataView As DataView)
                Dim WarehouseName As String = Warehouse_Get(oWarehouse.WarehouseID).Name.Trim

                _database.ClearParameter()
                _database.AddParameter("@WarehouseID", oWarehouse.WarehouseID, True)
                AddWarehouseParametersToCommand(oWarehouse)
                _database.ExecuteSP_NO_Return("Warehouse_Update")

                For Each row As DataRow In DB.Van.Van_GetAll(False).Table.Rows
                    If Sys.Helper.MakeStringValid(row.Item("Registration")).Split("*")(1).Trim = WarehouseName Then
                        Dim oVan As New Vans.Van

                        oVan.SetVanID = row.Item("VanID")
                        oVan.SetRegistration = Sys.Helper.MakeStringValid(row.Item("Registration")).Split("*")(0).Trim & " * " & oWarehouse.Name
                        oVan.SetNotes = row.Item("Notes")
                        oVan.InsuranceDue = row.Item("InsuranceDue")
                        oVan.MOTDue = row.Item("MOTDue")
                        oVan.TaxDue = row.Item("TaxDue")
                        oVan.ServiceDue = row.Item("ServiceDue")
                        oVan.SetMileageLimit = row.Item("MileageLimit")
                        oVan.SetPeriodValue = row.Item("PeriodValue")
                        oVan.SetPeriodType = row.Item("PeriodType")

                        DB.Van.Update(oVan, LocationsDataView, True)
                    End If
                Next
            End Sub

            Private Sub AddWarehouseParametersToCommand(ByRef oWarehouse As Warehouse)
                With _database
                    .AddParameter("@Name", oWarehouse.Name, True)
                    .AddParameter("@Size", oWarehouse.Size, True)
                    .AddParameter("@Notes", oWarehouse.Notes, True)
                    .AddParameter("@Address1", oWarehouse.Address1, True)
                    .AddParameter("@Address2", oWarehouse.Address2, True)
                    .AddParameter("@Address3", oWarehouse.Address3, True)
                    .AddParameter("@Address4", oWarehouse.Address4, True)
                    .AddParameter("@Address5", oWarehouse.Address5, True)
                    .AddParameter("@Postcode", oWarehouse.Postcode, True)
                    .AddParameter("@TelephoneNum", oWarehouse.TelephoneNum, True)
                    .AddParameter("@FaxNum", oWarehouse.FaxNum, True)
                    .AddParameter("@EmailAddress", oWarehouse.EmailAddress, True)
                    .AddParameter("@Active", oWarehouse.Active, True)
                End With
            End Sub

            Public Function Warehouse_GetWithProduct(ByVal ProductID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@ProductID", ProductID)

                'Get the datatable from the database store in dt
                Return New DataView(_database.ExecuteSP_DataTable("Warehouse_GetWithProduct"))
            End Function

            Public Function Warehouse_GetWithProduct(ByVal ProductID As Integer, ByVal trans As SqlClient.SqlTransaction) As DataView

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_GetWithProduct"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@ProductID", ProductID)

                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                Return New DataView(dt)

            End Function

            Public Function Warehouse_GetWithPart(ByVal PartID As Integer, ByVal trans As SqlClient.SqlTransaction) As DataView

                Dim Command As SqlClient.SqlCommand = New SqlClient.SqlCommand

                Command.CommandText = "Warehouse_GetWithPart"
                Command.CommandType = CommandType.StoredProcedure
                Command.Transaction = trans
                Command.Connection = trans.Connection

                Command.Parameters.AddWithValue("@PartID", PartID)

                Dim Adapter As New SqlClient.SqlDataAdapter(Command)
                Dim returnDS As New DataSet
                Adapter.Fill(returnDS)

                Dim dt As DataTable = returnDS.Tables(0)

                Return New DataView(dt)

            End Function

            Public Function Warehouse_GetWithPart(ByVal PartID As Integer) As DataView
                _database.ClearParameter()
                _database.AddParameter("@PartID", PartID)

                'Get the datatable from the database store in dt
                Return New DataView(_database.ExecuteSP_DataTable("Warehouse_GetWithPart"))
            End Function

#End Region

        End Class

    End Namespace

End Namespace

File: src/Application Specific/Classes/Entity Collections/Warehouse/WarehouseSQL.cs
using System.Data;
using System.Data.SqlClient;
using Microsoft.VisualBasic.CompilerServices;

namespace FSM.Entity
{
    namespace Warehouses
    {
        public class WarehouseSQL
        {
            private Sys.Database _database;

            public WarehouseSQL(Sys.Database database)
            {
                _database = database;
            }

            /* TODO ERROR: Skipped RegionDirectiveTrivia */

            public void Delete(int WarehouseID)
            {
                string WarehouseName = Warehouse_Get(WarehouseID).Name.Trim();
                _database.ClearParameter();
                _database.AddParameter("@WarehouseID", WarehouseID, true);
                _database.ExecuteSP_NO_Return("Warehouse_Delete");
                foreach (DataRow row in App.DB.Van.Van_GetAll(false).Table.Rows)
                {
                    if ((Sys.Helper.MakeStringValid(row["Registration"]).Split('*')[1].Trim() ?? "") == (WarehouseName ?? ""))
                    {
                        App.DB.Van.Delete(Conversions.ToInteger(row["VanID"]), true);
                    }
                }
            }

            public DataView Warehouse_GetDV(int WarehouseID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_Get";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@WarehouseID", WarehouseID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public DataView Warehouse_GetDV(int WarehouseID)
            {
                _database.ClearParameter();
                _database.AddParameter("@WarehouseID", WarehouseID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Warehouse_Get");
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public Warehouse Warehouse_Get(int WarehouseID)
            {
                _database.ClearParameter();
                _database.AddParameter("@WarehouseID", WarehouseID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Warehouse_Get");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oWarehouse = new Warehouse();
                        oWarehouse.IgnoreExceptionsOnSetMethods = true;
                        oWarehouse.SetWarehouseID = dt.Rows[0]["WarehouseID"];
                        oWarehouse.SetName = dt.Rows[0]["warehouseName"];
                        oWarehouse.SetSize = dt.Rows[0]["Size"];
                        oWarehouse.SetNotes = dt.Rows[0]["Notes"];
                        oWarehouse.SetAddress1 = dt.Rows[0]["Address1"];
                        oWarehouse.SetAddress2 = dt.Rows[0]["Address2"];
                        oWarehouse.SetAddress3 = dt.Rows[0]["Address3"];
                        oWarehouse.SetAddress4 = dt.Rows[0]["Address4"];
                        oWarehouse.SetAddress5 = dt.Rows[0]["Address5"];
                        oWarehouse.SetPostcode = dt.Rows[0]["Postcode"];
                        oWarehouse.SetTelephoneNum = dt.Rows[0]["TelephoneNum"];
                        oWarehouse.SetFaxNum = dt.Rows[0]["FaxNum"];
                        oWarehouse.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        oWarehouse.SetActive = dt.Rows[0]["Active"];
                        oWarehouse.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oWarehouse.Exists = true;
                        return oWarehouse;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public Warehouse Warehouse_Get(int WarehouseID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_Get";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@WarehouseID", WarehouseID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oWarehouse = new Warehouse();
                        oWarehouse.IgnoreExceptionsOnSetMethods = true;
                        oWarehouse.SetWarehouseID = dt.Rows[0]["WarehouseID"];
                        oWarehouse.SetName = dt.Rows[0]["warehouseName"];
                        oWarehouse.SetSize = dt.Rows[0]["Size"];
                        oWarehouse.SetNotes = dt.Rows[0]["Notes"];
                        oWarehouse.SetAddress1 = dt.Rows[0]["Address1"];
                        oWarehouse.SetAddress2 = dt.Rows[0]["Address2"];
                        oWarehouse.SetAddress3 = dt.Rows[0]["Address3"];
                        oWarehouse.SetAddress4 = dt.Rows[0]["Address4"];
                        oWarehouse.SetAddress5 = dt.Rows[0]["Address5"];
                        oWarehouse.SetPostcode = dt.Rows[0]["Postcode"];
                        oWarehouse.SetTelephoneNum = dt.Rows[0]["TelephoneNum"];
                        oWarehouse.SetFaxNum = dt.Rows[0]["FaxNum"];
                        oWarehouse.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        oWarehouse.SetActive = dt.Rows[0]["Active"];
                        oWarehouse.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oWarehouse.Exists = true;
                        return oWarehouse;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public Warehouse Warehouse_GetByLocationID(int LocationID)
            {
                _database.ClearParameter();
                _database.AddParameter("@LocationID", LocationID);

                // Get the datatable from the database store in dt
                var dt = _database.ExecuteSP_DataTable("Warehouse_GetByLocationID");
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oWarehouse = new Warehouse();
                        oWarehouse.IgnoreExceptionsOnSetMethods = true;
                        oWarehouse.SetWarehouseID = dt.Rows[0]["WarehouseID"];
                        oWarehouse.SetName = dt.Rows[0]["warehouseName"];
                        oWarehouse.SetSize = dt.Rows[0]["Size"];
                        oWarehouse.SetNotes = dt.Rows[0]["Notes"];
                        oWarehouse.SetAddress1 = dt.Rows[0]["Address1"];
                        oWarehouse.SetAddress2 = dt.Rows[0]["Address2"];
                        oWarehouse.SetAddress3 = dt.Rows[0]["Address3"];
                        oWarehouse.SetAddress4 = dt.Rows[0]["Address4"];
                        oWarehouse.SetAddress5 = dt.Rows[0]["Address5"];
                        oWarehouse.SetPostcode = dt.Rows[0]["Postcode"];
                        oWarehouse.SetTelephoneNum = dt.Rows[0]["TelephoneNum"];
                        oWarehouse.SetFaxNum = dt.Rows[0]["FaxNum"];
                        oWarehouse.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        oWarehouse.SetActive = dt.Rows[0]["Active"];
                        oWarehouse.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oWarehouse.Exists = true;
                        return oWarehouse;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public Warehouse Warehouse_GetByLocationID(int LocationID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_GetByLocationID";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@LocationID", LocationID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                if (dt is object)
                {
                    if (dt.Rows.Count > 0)
                    {
                        var oWarehouse = new Warehouse();
                        oWarehouse.IgnoreExceptionsOnSetMethods = true;
                        oWarehouse.SetWarehouseID = dt.Rows[0]["WarehouseID"];
                        oWarehouse.SetName = dt.Rows[0]["warehouseName"];
                        oWarehouse.SetSize = dt.Rows[0]["Size"];
                        oWarehouse.SetNotes = dt.Rows[0]["Notes"];
                        oWarehouse.SetAddress1 = dt.Rows[0]["Address1"];
                        oWarehouse.SetAddress2 = dt.Rows[0]["Address2"];
                        oWarehouse.SetAddress3 = dt.Rows[0]["Address3"];
                        oWarehouse.SetAddress4 = dt.Rows[0]["Address4"];
                        oWarehouse.SetAddress5 = dt.Rows[0]["Address5"];
                        oWarehouse.SetPostcode = dt.Rows[0]["Postcode"];
                        oWarehouse.SetTelephoneNum = dt.Rows[0]["TelephoneNum"];
                        oWarehouse.SetFaxNum = dt.Rows[0]["FaxNum"];
                        oWarehouse.SetEmailAddress = dt.Rows[0]["EmailAddress"];
                        oWarehouse.SetActive = dt.Rows[0]["Active"];
                        oWarehouse.SetDeleted = Conversions.ToBoolean(dt.Rows[0]["Deleted"]);
                        oWarehouse.Exists = true;
                        return oWarehouse;
                    }
                    else
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }

            public DataView Warehouse_GetAll()
            {
                _database.ClearParameter();
                var dt = _database.ExecuteSP_DataTable("Warehouse_GetAll");
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public DataView Warehouse_GetAll_For_Van2(int VanID)
            {
                _database.ClearParameter();
                _database.AddParameter("@VanID", VanID);
                var dt = _database.ExecuteSP_DataTable("Warehouse_GetAll_For_Van2");  // rd change
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public DataView Warehouse_GetAll(SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_GetAll";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public Warehouse Insert(Warehouse oWarehouse, DataView LocationsDataView)
            {
                _database.ClearParameter();
                AddWarehouseParametersToCommand(ref oWarehouse);
                oWarehouse.SetWarehouseID = Sys.Helper.MakeIntegerValid(_database.ExecuteSP_OBJECT("Warehouse_Insert"));
                oWarehouse.Exists = true;
                var oLocation = new Locationss.Locations();
                oLocation.SetDeleted = false;
                oLocation.SetVanID = null;
                oLocation.SetWarehouseID = oWarehouse.WarehouseID;
                App.DB.Location.Insert(oLocation);
                foreach (DataRow row in App.DB.Van.Van_GetAll(true).Table.Rows)
                {
                    var oVan = new Vans.Van();
                    oVan.SetRegistration = row["Registration"] + " * " + oWarehouse.Name;
                    oVan.SetNotes = row["Notes"];
                    oVan.InsuranceDue = Conversions.ToDate(row["InsuranceDue"]);
                    oVan.MOTDue = Conversions.ToDate(row["MOTDue"]);
                    oVan.TaxDue = Conversions.ToDate(row["TaxDue"]);
                    oVan.ServiceDue = Conversions.ToDate(row["ServiceDue"]);
                    oVan.SetMileageLimit = row["MileageLimit"];
                    oVan.SetPeriodValue = row["PeriodValue"];
                    oVan.SetPeriodType = row["PeriodType"];
                    App.DB.Van.Insert(oVan, "", null, true);
                }

                foreach (DataRow row in App.DB.Van.Van_GetAll(true).Table.Rows)
                {
                    foreach (DataRow locationRow in LocationsDataView.Table.Rows)
                    {
                        if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(row["Registration"], locationRow["Registration"], false)))
                        {
                            int locID = Conversions.ToInteger(App.DB.Van.Van_GetAll_For_Warehouse(oWarehouse.WarehouseID).Table.Select(Conversions.ToString("Registration = '" + row["Registration"] + "'"))[0]["LocationID"]);
                            App.DB.Location.Update(locID, Conversions.ToBoolean((!(bool)locationRow["Tick"])));
                        }
                    }
                }

                return oWarehouse;
            }

            public DataView Warehouse_Search(string criteria, string SearchOnField)
            {
                if (SearchOnField.Trim().Length > 0)
                {
                    criteria = "'%" + criteria + "%'";
                }

                _database.ClearParameter();
                _database.AddParameter("@SearchString", criteria, true);
                _database.AddParameter("@SearchOnField", SearchOnField, true);
                var dt = _database.ExecuteSP_DataTable("Warehouse_Search");
                dt.TableName = Sys.Enums.TableNames.tblWarehouse.ToString();
                return new DataView(dt);
            }

            public void Update(Warehouse oWarehouse, DataView LocationsDataView)
            {
                string WarehouseName = Warehouse_Get(oWarehouse.WarehouseID).Name.Trim();
                _database.ClearParameter();
                _database.AddParameter("@WarehouseID", oWarehouse.WarehouseID, true);
                AddWarehouseParametersToCommand(ref oWarehouse);
                _database.ExecuteSP_NO_Return("Warehouse_Update");
                foreach (DataRow row in App.DB.Van.Van_GetAll(false).Table.Rows)
                {
                    if ((Sys.Helper.MakeStringValid(row["Registration"]).Split('*')[1].Trim() ?? "") == (WarehouseName ?? ""))
                    {
                        var oVan = new Vans.Van();
                        oVan.SetVanID = row["VanID"];
                        oVan.SetRegistration = Sys.Helper.MakeStringValid(row["Registration"]).Split('*')[0].Trim() + " * " + oWarehouse.Name;
                        oVan.SetNotes = row["Notes"];
                        oVan.InsuranceDue = Conversions.ToDate(row["InsuranceDue"]);
                        oVan.MOTDue = Conversions.ToDate(row["MOTDue"]);
                        oVan.TaxDue = Conversions.ToDate(row["TaxDue"]);
                        oVan.ServiceDue = Conversions.ToDate(row["ServiceDue"]);
                        oVan.SetMileageLimit = row["MileageLimit"];
                        oVan.SetPeriodValue = row["PeriodValue"];
                        oVan.SetPeriodType = row["PeriodType"];
                        App.DB.Van.Update(oVan, LocationsDataView, true);
                    }
                }
            }

            private void AddWarehouseParametersToCommand(ref Warehouse oWarehouse)
            {
                {
                    var withBlock = _database;
                    withBlock.AddParameter("@Name", oWarehouse.Name, true);
                    withBlock.AddParameter("@Size", oWarehouse.Size, true);
                    withBlock.AddParameter("@Notes", oWarehouse.Notes, true);
                    withBlock.AddParameter("@Address1", oWarehouse.Address1, true);
                    withBlock.AddParameter("@Address2", oWarehouse.Address2, true);
                    withBlock.AddParameter("@Address3", oWarehouse.Address3, true);
                    withBlock.AddParameter("@Address4", oWarehouse.Address4, true);
                    withBlock.AddParameter("@Address5", oWarehouse.Address5, true);
                    withBlock.AddParameter("@Postcode", oWarehouse.Postcode, true);
                    withBlock.AddParameter("@TelephoneNum", oWarehouse.TelephoneNum, true);
                    withBlock.AddParameter("@FaxNum", oWarehouse.FaxNum, true);
                    withBlock.AddParameter("@EmailAddress", oWarehouse.EmailAddress, true);
                    withBlock.AddParameter("@Active", oWarehouse.Active, true);
                }
            }

            public DataView Warehouse_GetWithProduct(int ProductID)
            {
                _database.ClearParameter();
                _database.AddParameter("@ProductID", ProductID);

                // Get the datatable from the database store in dt
                return new DataView(_database.ExecuteSP_DataTable("Warehouse_GetWithProduct"));
            }

            public DataView Warehouse_GetWithProduct(int ProductID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_GetWithProduct";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@ProductID", ProductID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                return new DataView(dt);
            }

            public DataView Warehouse_GetWithPart(int PartID, SqlTransaction trans)
            {
                var Command = new SqlCommand();
                Command.CommandText = "Warehouse_GetWithPart";
                Command.CommandType = CommandType.StoredProcedure;
                Command.Transaction = trans;
                Command.Connection = trans.Connection;
                Command.Parameters.AddWithValue("@PartID", PartID);
                var Adapter = new SqlDataAdapter(Command);
                var returnDS = new DataSet();
                Adapter.Fill(returnDS);
                var dt = returnDS.Tables[0];
                return new DataView(dt);
            }

            public DataView Warehouse_GetWithPart(int PartID)
            {
                _database.ClearParameter();
                _database.AddParameter("@PartID", PartID);

                // Get the datatable from the database store in dt
                return new DataView(_database.ExecuteSP_DataTable("Warehouse_GetWithPart"));
            }

            /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        }
    }
}